### ISSUE TITLE
[BUG] Even when zoomed to 100%, the image is still compressed by the algorithm.

### ISSUE THREAD

**Issue by signin404** [justnullname/QuickView#239]
**Describe the bug**
Even when the zoom level is switched back to 100%, the image is still subject to algorithmic compression.
You have to manually drag the image to the bottom-left edge before you can actually switch to a 100% display without any algorithm-based interpolation.

**To Reproduce**
1. Open any image larger than the screen size.
2. Enable Lock window and Keep window size on navigation
3. Toggle zoom to non-100%.
4. Switch the zoom level to 100%.
5. At this point, even though the UI displays 100% zoom, the image quality is still compressed by the algorithm
6. When the image is moved to the bottom-left corner, the blurriness caused by the algorithm compression immediately disappears, allowing the image to truly switch to 100% display.

**Expected behavior**
Immediately stops any algorithmic interpolation when scaling is switched to 100%.

**Screenshots**

**Environment Information (CRITICAL)**
 - **QuickView Version**: v6.29.0.0
 - **OS Version**: Windows 10 21H2 19044.2006
 - **CPU**: 13600KF
 - **GPU**: 4070Ti
 - **RAM**: 32GB
 - **Display**: 1920x1080 100% Scale

**Image Information**
 - **Format**: JPEG
 - **Resolution**: 2609x1265
 - **File Size**: 1.82MB

**Additional context**
[QuickView.ini.txt](https://github.com/user-attachments/files/30828481/QuickView.ini.txt)
Recorded video: https://gofile.io/d/vURL4A
Test image: https://gofile.io/d/yULEdl

Can we add the feature of panning images with right-click drag?
Is it possible to disable the save dialog when switching to the next image after rotating an image or when exiting the program? I will never use it.

Why is there a constant CPU overhead even when the program is idle after rendering is complete?
Debugging the HUD: FPS still fluctuates even when idle.
Does this mean the program is still rendering and outputting new frames even when idle?
If so, can you optimize it so that it doesn't output new frames after rendering is complete to reduce the idle CPU overhead (Cycles delta) to the same 0 as MassiGra?

<img width="987" height="262" alt="Image" src="https://github.com/user-attachments/assets/070c9366-06dc-412a-b8e8-52da13629fa4" />

Would you be interested in adding the ability to interpolate using AI models?
For example, if the zoom level doesn’t change within 10 seconds, the displayed image will be replaced with one interpolated using AI.
When performing rapid zooming, traditional algorithms are still used to maintain real-time performance.
Model and dependency files can be large
QuickView can simply provide an interface for users to download the model and dependency files themselves if they need to.

**Comment by signin404** [comment]
I use [System Informer](https://github.com/winsiderss/systeminformer) to track resource usage.

AI interpolation may be able to call NVDIA VSR directly to avoid downloading huge dependencies and model files?

**Comment by cmixed** [comment]
Although I personally use an NVIDIA GPU, QuickView should remain accessible to everyone—including AMD and Intel GPU users. Adding features that rely on vendor-specific tech (like calling NVIDIA VSR directly) would create a significant maintenance burden and lead to a fragmented experience across different hardware. Supporting all GPU vendors properly is already challenging enough without adding AI model deployment into the mix.

I think keeping QuickView focused on fast, reliable, and hardware-agnostic rendering with traditional methods is the safer bet for now.

**Comment by signin404** [comment]
The new AI interpolation feature is awesome!
What a feat that's never been achieved in any image viewer software!

Here are some optimization suggestions:

When the AI-interpolated image cache already exists.
The delay in switching between >100% and ≤100% is not necessary because the switching can be done quickly.
The delay can be set to a hardcoded value of 0/300/500ms, or an independent setting option can be added.
Instead of directly using the same trigger delay as automatic AI interpolation.

AI interpolation size limits the MP units used is hard to understand.
It is recommended to use the width and highly of custom input as units.
If both the width and highly input values are 0, then unlimited.
If either the width or the highly has only one input value of 0, use only the non-zero input value as the constraint.

It seems the model file isn't in a common .onnx or .pth format. How is this converted?

Version 6.46.0.6 still suffers from 100% zoom blurriness
https://gofile.io/d/2VpGOaFU

**Comment by justnullname** [comment]
Hey, really glad to hear you're enjoying the feature, and thanks for the great suggestions!

Regarding the model format: 

The main reason we didn’t go with `.onnx` or `.pth` directly comes down to **binary size and dependencies**. Supporting `.pth` requires a heavy LibTorch/Python environment, and `.onnx` would mean bundling ONNX Runtime and DirectML DLLs (easily 50–100MB+ of extra bloat). To keep QuickView instant-starting and lightweight, we use Tencent's **ncnn** framework with a native Vulkan backend. Models are stored as `.param` (network graph) and `.bin` (weights), which keeps everything down to just a couple of megabytes without sacrificing GPU acceleration.

That said, I’m definitely planning to experiment with and bring in more powerful models down the road.

If you’d like to convert your own models, the pipeline is basically `PyTorch -> ONNX -> ncnn`:

1. **Export from PyTorch to ONNX:**
   ```python
   import torch

   torch.onnx.export(
       model,
       dummy_input,
       "model.onnx",
       opset_version=11,
       input_names=["input"],
       output_names=["output"]
   )
   ```

2. **Convert to ncnn using the official tools:**
   Grab the prebuilt binaries from [Tencent/ncnn releases](https://github.com/Tencent/ncnn/releases), then run:
   ```bash
   onnx2ncnn.exe model.onnx model.param model.bin
   ```

3. **Optimize & FP16 compression (recommended):**
   ```bash
   ncnnoptimize.exe model.param model.bin model_opt.param model_opt.bin 65536
   ```

**A quick note on compatibility:**
Currently, classic CNN super-resolution architectures (like Real-ESRGAN Compact, Real-CUGAN, etc.) work best. Models with custom unsupported layers or heavy transformer/attention blocks might fail to convert or require custom layer implementations in ncnn. Also, make sure the model accepts a standard 3-channel RGB input (normalized to $[0, 1]$) with dynamic resolution.

By the way, I've just addressed your other feedback as well, the cache switching is now instant (0ms delay), and the trigger thresholds have been updated to width/height limits instead of MP. Appreciate the input!

**Comment by signin404** [comment]
It would be great if we could set different custom models for adaptive (2x/3x/4x) respectively.

It is recommended to add an adaptive (2x/4x) option.
Using a 4x model when scaling > 2x and < 4x.
Because 3x models are rare, and their effectiveness is not guaranteed.
The current quality of realesr-animevideov3-x3 is even inferior to realesr-animevideov3-x2+FSR.

I found a strange problem.
I replaced realesr-animevideov3-x2, but the 2x mode in Anime Fast (Auto 2x/3x/4x) doesn't seem to use it.
Only when set to Anime Fast 2x will it be called.
Is this because it always calls realesr-animevideov3-x4 and adjusts the scaling factor via input parameters?

Conversion of onnx to ncnn using Google Colab (if anyone needs it)

```Python
# ================= 1. 环境准备 (下载工具与依赖) =================
!pip install -q onnx

# 注入 libprotobuf23 并拉取官方 20230816 预编译工具 (已跑过可秒跳过)
!wget -q -O /content/libprotobuf23.deb http://archive.ubuntu.com/ubuntu/pool/main/p/protobuf/libprotobuf23_3.12.4-1ubuntu7_amd64.deb
!dpkg -x /content/libprotobuf23.deb /content/proto_pkg
!cp -P /content/proto_pkg/usr/lib/x86_64-linux-gnu/libprotobuf.so.23* /usr/lib/x86_64-linux-gnu/
!ldconfig
!rm -rf /content/proto_pkg /content/libprotobuf23.deb

!wget -q -O /content/ncnn.zip https://github.com/Tencent/ncnn/releases/download/20230816/ncnn-20230816-ubuntu-2204.zip
!unzip -q -o /content/ncnn.zip -d /content/ncnn_pkg
!find /content/ncnn_pkg -type f -name "onnx2ncnn" -exec cp {} /content/ \;
!chmod +x /content/onnx2ncnn
!rm -rf /content/ncnn_pkg /content/ncnn.zip

import os
import math
import onnx
from onnx import numpy_helper
import torch
from torch import nn
import torch.nn.functional as F

# ================= 1. 定义标准 SRVGGNetCompact 架构 =================
class SRVGGNetCompact(nn.Module):
    def __init__(self, num_in_ch=3, num_out_ch=3, num_feat=64, num_conv=16, upscale=4, act_type='prelu'):
        super().__init__()
        self.num_in_ch = num_in_ch
        self.num_out_ch = num_out_ch
        self.num_feat = num_feat
        self.num_conv = num_conv
        self.upscale = upscale

        self.body = nn.ModuleList()
        self.body.append(nn.Conv2d(num_in_ch, num_feat, 3, 1, 1))
        self.body.append(nn.PReLU(num_parameters=num_feat))

        for _ in range(num_conv):
            self.body.append(nn.Conv2d(num_feat, num_feat, 3, 1, 1))
            self.body.append(nn.PReLU(num_parameters=num_feat))

        self.body.append(nn.Conv2d(num_feat, num_out_ch * upscale * upscale, 3, 1, 1))
        self.upsampler = nn.PixelShuffle(upscale)

    def forward(self, x):
        out = x
        for i in range(len(self.body)):
            out = self.body[i](out)
        out = self.upsampler(out)
        base = F.interpolate(x, scale_factor=self.upscale, mode='nearest')
        out += base
        return out

# ================= 2. 从已有 ONNX 逆向提取权重并重建 .pth =================
input_onnx = "1.onnx"  # 替换为你手头会崩溃的 ONNX 文件名
recovered_pth = "recovered_model.pth"

print(f"正在分析 ONNX: {input_onnx} ...")
onnx_model = onnx.load(input_onnx)

# 提取 ONNX 中所有的权重 Tensor
inits = {init.name: torch.from_numpy(numpy_helper.to_array(init)) for init in onnx_model.graph.initializer}

# 抓取所有 Conv 和 PRelu 算子节点
conv_nodes = [n for n in onnx_model.graph.node if n.op_type == "Conv"]
prelu_nodes = [n for n in onnx_model.graph.node if n.op_type == "PRelu"]

# 自动推断超分倍数与卷积层数
last_conv_weight = inits[conv_nodes[-1].input[1]]
out_channels = last_conv_weight.shape[0]
upscale = int(math.isqrt(out_channels // 3))
num_conv = len(conv_nodes) - 2

print(f"成功识别模型参数 -> 放大倍率: {upscale}x, 主体卷积层数: {num_conv}")

# 组装为标准的 PyTorch state_dict
state_dict = {}

# 第一层 Conv + PRelu
state_dict['body.0.weight'] = inits[conv_nodes[0].input[1]]
if len(conv_nodes[0].input) > 2:
    state_dict['body.0.bias'] = inits[conv_nodes[0].input[2]]
state_dict['body.1.weight'] = inits[prelu_nodes[0].input[1]].squeeze()

# 中间层 Conv + PRelu
for i in range(num_conv):
    c_node = conv_nodes[1 + i]
    p_node = prelu_nodes[1 + i]
    state_dict[f'body.{2 + 2*i}.weight'] = inits[c_node.input[1]]
    if len(c_node.input) > 2:
        state_dict[f'body.{2 + 2*i}.bias'] = inits[c_node.input[2]]
    state_dict[f'body.{3 + 2*i}.weight'] = inits[p_node.input[1]].squeeze()

# 最后一层 Conv
last_c_node = conv_nodes[-1]
state_dict[f'body.{2 + 2*num_conv}.weight'] = inits[last_c_node.input[1]]
if len(last_c_node.input) > 2:
    state_dict[f'body.{2 + 2*num_conv}.bias'] = inits[last_c_node.input[2]]

# 保存为干净的 .pth
torch.save(state_dict, recovered_pth)
print(f"🎉 成功逆向还原为干净的 PyTorch 权重文件: {recovered_pth}")

# ================= 3. 重新标准导出 (绝无冗余算子与倍率丢失) =================
model = SRVGGNetCompact(num_in_ch=3, num_out_ch=3, num_feat=64, num_conv=num_conv, upscale=upscale, act_type='prelu')
model.load_state_dict(state_dict)
model.eval()

dummy_input = torch.randn(1, 3, 64, 64)

torch.onnx.export(
    model,
    dummy_input,
    "model_clean.onnx",
    opset_version=11,
    dynamo=False,
    input_names=["data"],       # 标准 data 输入
    output_names=["output"],     # 标准 output 输出
    dynamic_axes={
        "data": {0: "batch", 2: "h", 3: "w"},
        "output": {0: "batch", 2: "h", 3: "w"}
    }
)

# 转换生成最终的 NCNN 模型
!./onnx2ncnn model_clean.onnx model.param model.bin
!rm -f model_clean.onnx

print(f"\n🎉 重新编译完成！已生成全新的 model.param 和 model.bin，请下载放入软件测试。")
```
